Add custom slug in module block

Imran92 opened the pull request:
Started as a solution to https://github.com/Automattic/sensei/issues/5176 . Also allows adding custom slugs to modules from the module block in the course edit page.

### Changes proposed in this Pull Request

1. On the course edit page, the block settings for the module block will contain a Custom slug section to add a custom slug for the module.
2. If a custom slug is added for a module, it'll be shown below the title of the module so that it's easy to notice and understand that there's a custom slug for that module
3. If the course teacher is changed, the module with the custom slug will belong to the new teacher.

The following 2 are a bit technical:
4. As custom slugs may not have the teacher id prefixed with the slug, we've started keeping the teacher's id as metadata to the module term so that if we need the id of the teacher of the module later (like for showing the name of the module teacher), we can use that meta value.
5. If a module's slug is changed to a custom slug from a conventional slug from the modules page instead of the course block editor page, a hook action saves the teacher id from the slug to the module metadata before changing the slug and losing the data.

### Testing instructions

- In SenseiLMS->Courses, click on 'New Course'. Add a few module blocks.
- Select a module block, and go to the Block settings panel on the right. Add a custom slug.
- That slug will appear below the module title as well. Add custom slug for some, keep some unchanged
- Try saving the course.
- The slugs should persist.
- Go to SenseiLMS->Modules page, change the slug of some modules with custom slugs and some with conventional slug.
- Open the same course again, make sure the edited slugs are visible. Try changing some texts on the editor page and save again. make sure everything is working.
- Change the teacher for the course and save. Make sure the changed teacher appears for all modules correctly.
- Go to modules page again and change the slug of a module that belongs to a teacher. Change the slug so that it does not append the teacher's id anymore.
- Go to the modules page again, make sure that module is still showing the teacher's name.

### Screenshot / Video

https://user-images.githubusercontent.com/6820724/171064046-5ca8eff5-26b5-4b7a-b217-9e06a7750f83.mov

Commit:
Add custom slug option in module block setting panel

Commit:
Show slug in block only if custom, pass slug value to setting

Commit:
Add slug and teacher id data in block json

Commit:
Extract teacher id and slug from module block

Commit:
Add teacher id to module block

This helps trigger a save of blocks after teacher is changed

Commit:
adjust custom slug css in module block

Commit:
Send teacher id and custom slug from api for module

Commit:
Try saving teacher id to meta if teacher slug is manually changed

Commit:
Get module author by meta if meta is available

Commit:
Save teacher to meta when updating teacher

Commit:
Save custom slug while storing structure

Commit:
Fix test by adding author to module meta

Commit:
Move teacher meta saving method to teacher class

Commit:
Test for adding teacher to meta from slug

Commit:
Add test for custom slug save

Commit:
Test custom slug show in module block

Commit:
Add private tag for hook

Commit:
Prevent event from binding multiple times

m1r0 commented:
Hey, I've noticed that the modules are duplicated when you edit the custom slug.

https://user-images.githubusercontent.com/1612178/172154937-c520e627-8dee-46db-ad46-6d395eec8da2.mp4

Another concern is that we currently allow modules to be owned by multiple courses which could be owned by multiple users (teachers/admins). Here's an example of what happens when you have a module owned by 2 courses that are owned by 2 different users (the owner of the last updated course wins):

https://user-images.githubusercontent.com/1612178/172157915-0ae0e9de-66eb-4a93-80df-00f96b2764c7.mp4

---

My opinion is that the old modules ownership logic should be fundamentally changed (if possible) and cleaned up to avoid future bugs. 
What do you think of not adding more fixes on top of it like the "custom slug" and trying to clean it up?

After analyzing the old code I found that it works like this (please correct me if I'm wrong):

* Modules have a one-to-one relation to teachers.
* To make the relation, the teacher id is added as a prefix to the slug.
* Because it's easy for users to edit the slug and break the relation, modules are duplicated when that happens.

I've noticed that you also landed on the same conclusion and you are using meta relation instead of slug relation. 🎉 🙂 🎉 

Maybe my approach would be something like the following:

1. Create a migration script that loops through the modules, gets the teacher id from the slug, and saves it as meta.
2. Take the meta relation a step further. If we have the ability to assign modules to multiple courses, then it's logical that the authors of those courses be able to edit those modules, right? What do you think of allowing a module to be owned by multiple teachers? Maybe the term could have one meta per author, for example `module_author_420` (meta keys are indexed so this will be very performant).
3. Clean up the teacher-slug-id logic and just use the meta.
4. Clean up the module duplication logic.

I don't have that much context about the domain, so I'm really curious what you think. 🙂

Imran92 commented:
Thanks a lot for the review @m1r0 , especially for taking the time to look into it in depth. As you've mentioned, we use slugs to allow teachers own the module they are teaching irrespective of course. So even if the teacher of a course is changed, the modules a teacher was teaching, still belongs to that teacher. Instead of changing the old module's teacher, we create new modules with new slugs but same titles that belongs to the new teacher. This all is maintained by system generated slugs, so I think the general assumption when developing modules was slugs will be system-generated for modules.

Once the user opts in (opt in because this is not a default behavior) to use a custom slug, and as till now we only allow one teacher per module (and we also show the custom slug on the module block so the user does not forget that a custom slug is being used for the module), when the user changes the teacher of the course, the expected outcome till now is that the module that has the user defined slug, should have the newly assigned teacher. Also with custom slug, we kinda had to use meta, but as you said, feels like patches. I was tempted to get rid of the slug based behaviors as you've said, but didn't as custom slug wasn't a requirement that I've come across too many times and to didn't want to risk breaking the current behavior that most users need, as it would require a lot of testing and covering a lot of cases to make sure nothing broke.

Having multiple teachers for a single module will definitely be the best solution in this case, I believe we've received some requests for this as well. But I think that will have a different scope, we can create another issue for that and take care separately maybe?

Commit:
Delete previous module if not in use after slug change

m1r0 commented:
> I was tempted to get rid of the slug based behaviors as you've said, but didn't as custom slug wasn't a requirement that I've come across too many times and to didn't want to risk breaking the current behavior that most users need, as it would require a lot of testing and covering a lot of cases to make sure nothing broke.

Yeah, I get your point. I have the same fear. 😄

After some more research, I think I know why modules are duplicated now - a teacher could use the same module on multiple courses. In other words - modules are re-used between courses. This complicates things a lot. 😄 My best guess is that we are trying to avoid duplicates in the database, I don't see another reason for it.

Here's a bug I've found related to that:

https://user-images.githubusercontent.com/1612178/172222439-088b97c9-9c46-49e1-9d99-5f89fa428728.mp4

> But I think that will have a different scope, we can create another issue for that and take care separately maybe?

Yeah, I agree. We should think of a proper solution once we have the time for it, so I'm ok with the current solution. 🙂

Commit:
Save teacher meta only when the teacher is not an admin

Commit:
Try managing one author one module per course by reusing

Commit:
Update tests to follow current code change for teacher meta save

Imran92 commented:
Thanks again for the updated review @m1r0 ☺️ 

> After some more research, I think I know why modules are duplicated now - a teacher could use the same module on multiple courses. In other words - modules are re-used between courses. This complicates things a lot. 😄 My best guess is that we are trying to avoid duplicates in the database, I don't see another reason for it.

Probably, I think so too. Maybe it will be worth looking at the whole requirement for modules and changing the flow and behavior to remove the complications later.

Thanks for pointing out the bug, it was an interesting one. Changed the behavior a bit to make it more consistent with the behavior of module [here](https://github.com/Automattic/sensei/pull/5207/commits/ce396db1b3325a7ff8050bdcd71f542ba1cc6058), hopefully this will solve the issue you faced. Can you kindly take a look again?

m1r0 commented:
I'm not sure if this is expected or not, but if you have 2 modules and use the same custom slug, the modules get merged. I did a few more tests and this allows for a teacher to overwrite modules that he doesn't own. Maybe there should be an error saying that slug is already in use?

m1r0 commented on includes/class-sensei-course-structure.php:
```diff
@@ -300,13 +304,46 @@ public function save( array $raw_structure ) {
-				$existing_module_id = $this->get_existing_module( $item['title'] );
-				$module_id          = $existing_module_id ? $existing_module_id : $this->create_module( $item );
+				// First try to get an existing module directly by the provided slug.
+				$existing_module = get_term_by( 'slug', $slug, 'module' );
+				// If not found, try to find if this teacher already has a module for this, with the default slug.
+				if ( ! $existing_module && $slug !== $default_slug ) {
+					$existing_module = get_term_by( 'slug', $default_slug, 'module' );
+				}
+				// If not found, try to find module for this teacher using term-meta.
+				if ( ! $existing_module && ! user_can( $author, 'manage_options' ) ) {
+					$modules = get_terms(
+						array(
+							'taxonomy'   => 'module',
+							'name'       => $item['title'],
+							'hide_empty' => false,
+							'meta_query' => array( // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query -- Only executed while saving, not too many data possibly.
+								array(
+									'key'     => 'module_author',
+									'value'   => $author,
+									'compare' => '=',
+								),
+							),
+						)
+					);
+					if ( ! empty( $modules ) ) {
+						$existing_module = $modules[0];
+					}
```
What do you think about extracting the "get existing module" logic to a separate method?

m1r0 commented on includes/class-sensei-teacher.php:
```diff
@@ -1649,4 +1654,29 @@ public static function remove_course_meta_on_teacher_archive() {
+		$term       = get_term( $term_id, 'module' );
+		$split_slug = explode( '-', $term->slug );
+
+		if ( count( $split_slug ) > 1 && is_numeric( $split_slug[0] ) ) {
```
Hm, will this logic fail if I name my module `2 Hours`? I guess this is an issue with the previous implementation. 😄

Imran92 replied:
Yeah 😄 That's how it has been. I'm afraid it's a bit hard to change at this point. We can think about changing it later, maybe when we work on allowing multiple teachers teach a single module

m1r0 commented:
Could you improve the name of the PR so it would read better in the changelog?

Commit:
Dont let use slug if already in use

Imran92 commented:
> I'm not sure if this is expected or not, but if you have 2 modules and use the same custom slug, the modules get merged. I did a few more tests and this allows for a teacher to overwrite modules that he doesn't own. Maybe there should be an error saying that slug is already in use?

Good point. I actually noticed it before, but thought as the user has opted in to use custom slug which has the highest priority, it should be the user's discretion. But now I think it may create more problems. So I've updated it as per your suggestion. There is a few things we consider regarding this -

- If the custom slug is already in use by another module, but it belongs to the same teacher, it'll get merged. It seems like the more rational way.
- If a custom slug belongs to another module by another teacher, but the module is not being used in any course, in that case the module with the custom slug will get merged with the existing module with custom slug and the resultant module will belong to the new teacher, only if the requester current user is an admin. Because the existing module belongs to another user, while an admin should be able to change the ownership a module, a teacher shouldn't be able to take ownership of another teacher's module no matter the module is being used by any course or not.
- If a custom slug belongs to another module which is being used in a course, you can not use that slug for any other teacher even if you are an admin. Because one module can't belong to two teachers.
- If two courses shares a common module with a custom slug and both the courses are taught by the same teacher, you can not change the teacher of one course. Because then the module with the custom slug will have to belong to two teachers.

There are various intricate scenarios here, I tried to handle all the above and a few others. Please let me know if you come up with any other scenario that you think needs to be covered :) Thanks again for reviewing it. I've updated the code as per your suggestions here https://github.com/Automattic/sensei/pull/5207/commits/a1a85e06eda81d46aef4e6af0b8ae860c6804b7c, can you kindly take a look again?

Commit:
Update check for finding module belonging to other user

m1r0 reviewed (commented):
Looks good and thanks for the explanation!

Do you think the new logic needs tests?

gabrielcaires commented:
@Imran92  @m1r0 
I think this [bug related]( https://github.com/Automattic/sensei/issues/5346) to this discussion
When the module title is changed, the backend creates a new module and returns a new module id.
The front-end is using it to update the store, but when the front receives a new module id, the entire module configuration (colors, border settings, etc...) is reset.

These lines are doing it: 

https://github.com/Automattic/sensei/pull/5207/files#diff-79524718ef30eb7796ccc6a458f0e98b24af517d330d295f39e3c072675474aeR312-R319

Commit:
Merge branch 'trunk' into add/custom-slug-in-module-block

Imran92 committed:
Add tests for module and course structure saving with and without custom slug

Imran92 commented:
> Looks good and thanks for the explanation!
> 
> Do you think the new logic needs tests?

Hi @m1r0 👋 I've added some tests for different logic regarding the structure save, hopefully, most of the cases are covered. Can you kindly check? Thanks https://github.com/Automattic/sensei/pull/5207/commits/8b4fdacc362250954950000efc7e07c8c467a1e0

m1r0 commented:
For some reason, [this CI run](https://github.com/Automattic/sensei/runs/7408367365?check_suite_focus=true) is not passing. Otherwise, the tests look good!

Imran92 committed:
Update test to pass in wp 5.7

Imran92 commented:
> For some reason, [this CI run](https://github.com/Automattic/sensei/runs/7408367365?check_suite_focus=true) is not passing. Otherwise, the tests look good!

The CI fail was a really tricky one :D was only failing for wp 5.7, Fixed here https://github.com/Automattic/sensei/pull/5207/commits/b143681a047bb71c23636321fbf9c6ad60d31f06

m1r0 reviewed (approved):
Nice!
